### DISCUSSION TITLE
Useful Debian 12 VM Commands

### DISCUSSION THREAD

**Discussion by tteck** [tteck/Proxmox#1988] [Announcements]
### Select xterm.js under the Console pull down for copy/paste functions
![Screenshot 2023-11-16 9 48 54 AM](https://github.com/tteck/Proxmox/assets/21342670/2a01de8c-4f08-41ef-a815-c2a807a3dfba)

## Get SSH Going
```
passwd root
```
```
sed -i -e 's/#PermitRootLogin prohibit-password/PermitRootLogin yes/g' -e 's/^PasswordAuthentication.*/PasswordAuthentication yes/' /etc/ssh/sshd_config
ssh-keygen -A
systemctl restart sshd
```

### Hardware > Hard Disk (scsi0) > Disk Action > Resize
![Screenshot 2023-11-16 11 56 40 AM](https://github.com/tteck/Proxmox/assets/21342670/4b9d02aa-e5a6-4270-a442-3340fafca4ed)

## Expand VM Disk using parted (/dev/sda1)
```
apt-get update
apt-get install -y parted
```

`parted /dev/sda`
`resizepart 1`
Fix/Ignore? `Fix`
Partition number? `1`
Yes/No? `Yes`
End?  [2146MB]? `-0`
(parted) `quit`
(reboot if not going further)

## Add Guest Agent
```
apt-get install qemu-guest-agent
```

## Change Hostname (replace "docker-vm" with whatever you want)
```
hostnamectl set-hostname docker-vm
sed -i '2i127.0.1.1       docker-vm' /etc/hosts
```
```
reboot
```

## Install Docker
```
sh <(curl -sSL https://get.docker.com)
```
## Install Docker Compose
```
LATEST=$(curl -sL https://api.github.com/repos/docker/compose/releases/latest | grep '"tag_name":' | cut -d'"' -f4)
DOCKER_CONFIG=${DOCKER_CONFIG:-$HOME/.docker}
mkdir -p $DOCKER_CONFIG/cli-plugins
curl -sSL https://github.com/docker/compose/releases/download/$LATEST/docker-compose-linux-x86_64 -o ~/.docker/cli-plugins/docker-compose
chmod +x $DOCKER_CONFIG/cli-plugins/docker-compose
docker compose version
```
## Add Dockge
* Install Docker Compose first
```
mkdir -p /opt/{dockge,stacks}
wget -q -O /opt/dockge/compose.yaml https://raw.githubusercontent.com/louislam/dockge/master/compose.yaml
cd /opt/dockge
docker compose up -d
```
Dockge GUI, **https:// IP:5001**
## Add Portainer
* Docker run
```
docker volume create portainer_data
docker run -d \
  -p 8000:8000 \
  -p 9443:9443 \
  --name=portainer \
  --restart=always \
  -v /var/run/docker.sock:/var/run/docker.sock \
  -v portainer_data:/data \
  portainer/portainer-ce:latest
```
Portainer GUI, **https:// IP:9443**

**Comment by click0** [comment]
`passwd root`
can be replaced with:
`echo -e "MyNewPasswd\nMyNewPasswd" | passwd root`
or
`echo root:MyNewPasswd| chpasswd`

**Reply by hollie** [reply]
Remember that this leaves the password in plain-text in the `~/.bash_history` file on disk.

**Comment by di5cord20** [comment]
@tteck - as a suggestion, linking to this announcement from the Helper Scripts page directly would make it easier for people to find/leverage (might reduce support questions in the future as well)

**Reply by tteck** [reply]
https://github.com/tteck/Proxmox/blob/48f59062462c0470d6b2027a228487bbd5d738e1/vm/debian-vm.sh#L423

I can do that...done

**Comment by nkwsy** [comment]
I think that the default size of the debian 12 VM is too small. I am unable to install parted in a 2gb VM

**Reply by scriptsandthings** [reply]
@nkwsy It works best for me by installing parted prior installing updates or packages.

`apt-get update && apt-get install parted -y`

Then, once that command is complete and parted is installed, follow the steps to expand the disk.
Once you've expanded the disk, immediately reboot the VM prior to doing any updates or other installations.

I've been updating the hostname, setting the root password, and configuring SSH prior to rebooting as that doesn't take up additional disk space. Def reboot if doing much else though.

**Reply by nkwsy** [reply]
Yes makes sense now. It might be nice if one could edit the size of the vm in the options. I normally install 'glances' on a new vm but I guess it has 950mb of python dependencies which clogged the system. 

I'll make sure to resize right away from now on

**Comment by andorardo** [comment]
These instructions are extremely helpful!

Question: How can I modify the network settings? (From DHCP to static.)

(I tried a few things but nothing stuck. I searched for docs online, but a lot of them point to /etc/network/interfaces, which is just not how the debian 12 vm image configures the interfaces.)

UPDATE with TL;DR: Debian uses netplan by default. See `/etc/netplan/90-default.yaml` instead.

**Reply by scriptsandthings** [reply]
I just had to do something similar for the AdBlockHome LXCs so Unbound would work properly. I left the LXC itself configured for DHCP, and assigned a static IP address reservation on my router for the LXC instead of a static IP configured in the LXC. I could change this, however I'm moving things between subnets and vlans reorganizing my network at the moment, so this has been working well for me.

Here's two guides that helped to keep Proxmox from overwriting /etc/hosts and /etc/resolv.conf that might help get you where you want to be.

To keep Proxmox from overriding resolv.conf: https://www.cyberciti.biz/faq/dhclient-etcresolvconf-hooks/
What worked for me on that page was following the part under **Method 2 Creating dhclient-script hooks** "_Create hook to avoid /etc/resolv.conf file update_"

And to keep Proxmox from messing with /etc/hosts and static IPs - https://pve.proxmox.com/pve-docs/chapter-pct.html#_guest_operating_system_configuration

On that page it discusses setting the **ostype** value manually to **'unmanaged'**. 

```
 ostype: <alpine | archlinux | centos | debian | devuan | fedora | gentoo | nixos | opensuse | ubuntu | unmanaged>

OS type: This is used to setup configuration inside the container, and corresponds to lxc setup scripts in /usr/share/lxc/config/<ostype>.common.conf. Value unmanaged can be used to skip and OS specific setup.
```

After LXC creation, you can go to /etc/pve/lxc/, find the **xxx.conf** file corrisponding to the LXC container ID you're working with and manually change the **'ostype'** value from **'debian'** to **'unmanaged'**. 

After making that edit, save your changes, reboot the LXC container if it was running, then you should be good to go :-)

**Reply by andorardo** [reply]
Thanks for the suggestion! I can definitely pin the IP address in the DHCP server, but since I have reserved IP ranged, I thought I would cut out the middle-man and set the IP address directly.

This must be doable, right? (I have an Alpine LXC container, and was able to set a static IP.)

FWIW, I had a look in `/etc/pve/qemu-server/108.conf`, but unfortunately there's no ostype in these VM setups.

**Reply by scriptsandthings** [reply]
Oh, I misread and forgot this was a thread on VMs, not LXC containers lol

You do set the IP inside of the VM at /etc/network/networkinterfaces 

There isn't a setting in proxmox for that with VMs. Just the MAC address and vlan tags are set in proxmox (more or less). 


https://www.cyberciti.biz/faq/add-configure-set-up-static-ip-address-on-debianlinux/

**Reply by andorardo** [reply]
Unfortunately, editing `/etc/network/interfaces` does *not* work. I tried all the standard Google searches, which is why I'm reaching out here :)
(The Debian 12 VM does not appear to use `networking.service`.)

`nmtui` does give the appearance of working, but the changes do not stick, for some reason :/

**Reply by scvnc** [reply]
I am experiencing the same thing after running the script today.

`systemctrl restart networking` results in systemctl erroring that the networking.service unit was not found.  something is fishy with the qemu image?  The Debian12 wiki suggests that we should have this per [`NetworkConfiguration -> Reinitialize new network setup`](https://wiki.debian.org/NetworkConfiguration)

**Reply by PerkySloth** [reply]
Got the same issue and found out that the system is using netplan, you have to modify the following file according to netplan syntax :

`/etc/netplan/90-default.yaml`

**Reply by jeremy-bepoix** [reply]
Change dhcp to static : 

edit `nano /etc/network/interfaces`

Modify this way as exemple _(must match your infrastucture)_

```
# This file describes the network interfaces available on your system
# and how to activate them. For more information, see interfaces(5).

source /etc/network/interfaces.d/*

# The loopback network interface
auto lo
iface lo inet loopback

auto ens18
# The primary network interface
allow-hotplug ens18
iface ens18 inet static
address 192.168.148.147/24
gateway 192.168.48.1
dns-nameservers 192.168.20.2
```

Finnaly restart service:
`systemctl restart networking`

**Reply by andorardo** [reply]
Please see above: editing `/etc/network/interfaces` does _not_ work.

See @PerkySloth 's answer.

**Reply by SysAdmln** [reply]
### 1. Configure IP Address and Default Route

Run the following commands to set the IP address and default gateway:

```bash
ip addr add 192.168.1.2 dev ens18
ip route add default via 192.168.1.1 dev ens18
```

**Note:** Replace `192.168.1.2` and `192.168.1.1` with your desired IP address and gateway if necessary.

### 2. Configure DNS

Add a DNS nameserver to the configuration:

```bash
echo "nameserver 8.8.8.8" >> /etc/resolv.conf
```

### 3. Remove Netplan and Install ifupdown

Remove `netplan.io` and install `ifupdown` for network management:

```bash
apt purge netplan.io
apt install ifupdown
```

### 4. Edit Network Interfaces Configuration

Open and edit the network interfaces configuration file:

```bash
nano /etc/network/interfaces
```

Add the following content to the file:

```bash
# Include configurations from /etc/network/interfaces.d/
source /etc/network/interfaces.d/*

# Loopback interface
auto lo
iface lo inet loopback

# Static IP configuration for ens18
auto ens18
allow-hotplug ens18
iface ens18 inet static
    address 192.168.1.2/24
    gateway 192.168.1.1
```

**Note:** Substitute `192.168.1.2` and `192.168.1.1` with your desired IP address and gateway.
```

**Reply by andorardo** [reply]
> `apt purge netplan.io && apt install ifupdown` edit `nano /etc/network/interfaces`

Interesting! So basically manually switch from `netplan` to `ifupdown`. 

Thanks for the tip!

PS: I'm only just realizing that Ubuntu/Debian et al. migrated to netplan [years ago](https://wiki.ubuntu.com/MigratingToNetplan#Rationale) 😮

**Reply by adrianxx85** [reply]
> ### 2. Configure DNS
> Add a DNS nameserver to the configuration:
> 
> ```shell
> echo "nameserver 8.8.8.8" >> /etc/resolv.conf



VM lose these after reboot -.-'

**Reply by daschmidt1994** [reply]
sudo nano /etc/NetworkManager/NetworkManager.conf

then 
[main]
dns=8.8.8.8

**Comment by Badn1nja** [comment]
Another easy way out of re-sizing the disk without parted is found here.. https://serverfault.com/a/994465

On a running system, run fdisk and carefully observe where the partition starts (p). Then remove it (d), and create (n) larger one on its place. Modern fdisk could ask about wiping signatures, if it will, say "no" (this is essential)! Pay special attention on its start, it must begin at exact same sector, but could have larger size. Exit fdisk writing changes (w).

+ reboot :)   (I had issues with lack of space because I rushed to download some tools.. unable to dl parted! )

**Comment by avluis** [comment]
With the source image for this install one can use the following set of commands, no need for parted:
```bash
growpart /dev/sda 1
resize2fs /dev/sda1
```
Done (no need for reboot).
Believe this will work for the Ubuntu VM script as well.
In case upstream nukes anything, `growpart` is part of the cloud-utils-package (`apt-get install -y cloud-utils`)

**Reply by derekslenk** [reply]
I wish this worked with CTs :( I created one too small (plex) and can't get the OS to know the disk has been resized, but since it's an LVM it doesn't register any disks, but the LG and PV don't show anything as well.

**Reply by tteck** [reply]
easy with a container

![Screenshot 2024-08-08 10 40 48 AM](https://github.com/user-attachments/assets/b0fda0ec-92c7-4281-a2b1-dc1c11ee50f9)

**Reply by derekslenk** [reply]
I thought the OS wasn't seeing it...but it was. I am....tired. Yup it was easy :)

**Comment by mrubiodev** [comment]
Change dhcp to static :
Install net-tools 
`apt-get install net-tools
`
look what your adapter is with
`ifconfig
`
create the following file with the name of the adapter
```
sudo nano /etc/systemd/network/10-<adapter>.network

```
and write information conection
```
[Match]
Name=ens18

[Network]
Address=<Address_device>/<subnet>
Gateway=<Address_gategay>
DNS=<Address>

```


for example 

```
sudo nano /etc/systemd/network/10-eth0.network


[Match]
Name=eth0

[Network]
Address=192.168.1.2/24
Gateway=192.168.1.1
DNS=1.1.1.1


```

and finaly
`sudo systemctl restart systemd-networkd`

**Reply by mupsje** [reply]
for vlan like me ;)

1. Install Required Tools:
Make sure you have net-tools and systemd-networkd installed (if not already).

Install net-tools
`apt-get install net-tools `
look what your adapter is with
`ifconfig `
create the following file with the name of the adapter

```
nano /etc/systemd/network/10-ens18.network
```

and write information conection (in my case) vlan15 
```
[Match]
Name=ens18

[Network]
VLAN=vlan15
```
2. Create VLAN NetDev File:
Create a `.netdev` file to define VLAN 15:
```
sudo nano /etc/systemd/network/vlan15.netdev
```
Add the following content::
```
[NetDev]
Name=vlan15
Kind=vlan

[VLAN]
Id=15
```
3. Create Network Configuration for VLAN 15:
Create the `.network` file to configure the VLAN interface with a static IP:

```
sudo nano /etc/systemd/network/10-vlan15.network
```
Add the following configuration:
```
[Match]
Name=vlan15

[Network]
Address=192.168.15.101/24
Gateway=192.168.15.1
DNS=192.168.15.1
```
4. Restart `systemd-networkd`:
Restart the network service to apply the changes:
```
sudo systemctl restart systemd-networkd
```
5. Bring Up VLAN Interface:
Manually bring up the VLAN interface (if not already up):
```
sudo ip link set vlan15 up
```

6. Enable systemd-networkd to Start at Boot:
Run the following command to ensure that `systemd-networkd` is enabled to start on boot:
```
sudo systemctl enable systemd-networkd
```
7. Reboot and Verify:
Reboot the system to check if the configuration persists:
```
sudo reboot
````

After reboot, verify the configuration:
```
ip addr show vlan15
```

**Comment by renz-id** [comment]
What to do if `sed -i -e 's/#PermitRootLogin prohibit-password/PermitRootLogin yes/g' -e 's/^PasswordAuthentication.*/PasswordAuthentication yes/' /etc/ssh/sshd_config`

return a `sed: can't read /etc/ssh/sshd_config: No such file or directory`?

**Reply by derekslenk** [reply]
I needed to install openssh-server first `apt-get install openssh-server`

**Reply by renz-id** [reply]
Yes, I forgot to update what solution I did for this particular issue.

**Reply by angelo-karta** [reply]
If anyone getting an error
`Package openssh-server is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source`

try `sudo apt-get update` first then `apt-get install openssh-server`

**Comment by romka777** [comment]
### Add Guest Agent

```
apt-get install qemu-guest-agent
systemctl enable qemu-guest-agent
systemctl start qemu-guest-agent
```

**Comment by derreisende77** [comment]
I had to `sudo apt install openssh-server` before ssh server was available in the VM.

**Comment by familie-braun** [comment]
Hello. After installing the Debian VM script, I called xterm.js. The console then opens with the message starting serial terminal on interface serial0. No matter what I enter, no command is executed. Can someone please help me?
<img width="1012" alt="Bildschirmfoto 2024-09-17 um 18 53 12" src="https://github.com/user-attachments/assets/a842e887-9031-4ea7-a40f-3af70ab7a123">
<img width="1055" alt="Bildschirmfoto 2024-09-17 um 18 54 01" src="https://github.com/user-attachments/assets/8322d91f-621d-4fdd-a63d-561749368411">
<img width="1322" alt="Bildschirmfoto 2024-09-17 um 18 54 52" src="https://github.com/user-attachments/assets/f83f22b6-6372-4cef-a497-369fd07447ef">
<img width="1605" alt="Bildschirmfoto 2024-09-17 um 18 55 09" src="https://github.com/user-attachments/assets/94659756-d5d3-4bbc-bce3-63a075d43576">

**Reply by saurori** [reply]
Worked after hitting "Enter" and it will prompt for `localhost login`

**Comment by ghost** [comment]
I couldn't update the installed virtual machine due to Debian using more space on /boot. I edited /etc/initramfs-tools/initramfs.conf by changing the MODULES=most to MODULES=dep.

**Comment by Lupu5R3x** [comment]
I typically doe's this:

`VMID=$(pvesh get /cluster/nextid) && bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/vm/debian-vm.sh)" && qm disk resize $VMID scsi0 60G` in proxmox shell, this auto expands the disk size upon creation.

Then I call this from my internal webserver: "ssh.sh"
```
#!/usr/bin/env bash
# Check if the script is run as root (or using sudo)
if [ "$(id -u)" -ne 0 ]; then
  echo "Please run this script as root or using sudo."
  exit 1
fi
# Check if openssh server is installed.
if command -v sshd &> /dev/null; then
    echo "openssh-server is installed."
else
    echo "openssh-server is not installed."
    echo "Installing openssh-server."
    apt-get update && apt-get install -y openssh-server
fi

sed -i -e 's/#PermitRootLogin prohibit-password/PermitRootLogin yes/g' -e 's/^PasswordAuthentication.*/PasswordAuthentication yes/' /etc/ssh/sshd_config
ssh-keygen -A
systemctl restart sshd

PERMIT_ROOT_LOGIN=$(grep -E '^PermitRootLogin' /etc/ssh/sshd_config)
PASSWORD_AUTH=$(grep -E '^PasswordAuthentication' /etc/ssh/sshd_config)
SSH_STATUS=$(systemctl is-active sshd)

if [[ "$PERMIT_ROOT_LOGIN" == "PermitRootLogin yes" ]] && [[ "$PASSWORD_AUTH" == "PasswordAuthentication yes" ]] && [[ "$SSH_STATUS" == "active" ]]; then
    echo "SSH configuration and service validated successfully."
else
    echo "There was an issue applying the SSH configuration changes."
    echo "Current settings:"
    echo "$PERMIT_ROOT_LOGIN"
    echo "$PASSWORD_AUTH"
    echo "SSH Service Status: $SSH_STATUS"
    exit 1
fi

echo 'Perfect'
```

And then I do call this: "expand_disk_size.sh"
```
#!/bin/bash

# This will check if parted is installed, if not it will install parted
# Then it will update the disk to use all available space.
# This script will only work if, debian is installed using
# the helper scripts from : https://tteck.github.io/Proxmox/#debian-12-vm

# I have only tested the script on Debian 12, installed in proxmox using, the above.

# Check if the script is run as root (or using sudo)
if [ "$(id -u)" -ne 0 ]; then
  echo "Please run this script as root or using sudo."
  exit 1
fi

# Check if parted is installed, and install it if not
if ! command -v parted &> /dev/null; then
  echo "parted is not installed. Installing it now..."
  
  # Install parted based on the package manager
  if command -v apt-get &> /dev/null; then
    apt-get update && apt-get install -y parted
  elif command -v yum &> /dev/null; then
    yum install -y parted
  else
    echo "Could not determine package manager. Please install 'parted' manually."
    exit 1
  fi
fi
printf 'fix\n1\nyes\n-0' | parted /dev/sda resizepart 1 ---pretend-input-tty
echo "We need to reboot, for the updates to take place!"
reboot
```

Finally I do a `apt-get install qemu-guest-agent -y && apt-get install -y mc` just to wrap things up :smile: 

But non of the above would be useful with out the cool helper scripts :1st_place_medal:
